### ISSUE TITLE
Embrace iterators and remove the `#[end]` variant?

### ISSUE THREAD

**Issue by maciejhirsz** [maciejhirsz/logos#105]
Since I've now implemented iterators for the next release, the question arises: should we drop the current `lex.token` + `lex.advance()` interface entirely, and use `Option<Token>` internally?

Given Rust optimizes nested enums, there doesn't seem to be much of a difference performance wise:

```rust
enum Token {
    End,
    Error,
    Other,
    Stuff,
    You,
    Might,
    Use,
}

#[no_mangle]
fn is_end_flat(token: Token) -> bool {
    match token {
        Token::End => true,
        _ => false,
    }
}

#[no_mangle]
fn is_end_option(token: Option<Token>) -> bool {
    token.is_none()
}
```

Produces ASM ([playground](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=df14b774541f74e6fd5e2ea74dd743d9)):

```asm
is_end_flat:
	testb	%dil, %dil
	sete	%al
	retq

is_end_option:
	cmpb	$7, %dil
	sete	%al
	retq
```

If `End` is moved to be the last variant (so that its discriminant value is non-zero), they become identical, spare `$6` instead of `$7`.

Lexer can work more or less the same as it does now, but instead of assigning `lex.token = Token::NewVariant` it could do `std::mem::replace(&mut lex.token, Some(Token::NewVariant))`, thus automatically returning `Option<Token>` for the previous token.

**Comment by maciejhirsz** [comment]
@ratmice: you had a use case for setting discriminant value 0 to the end variant, would removing it entirely in favor of `Option` break things for you?

**Comment by ratmice** [comment]
Thanks for the ping,

I'm currently using iterators for interfacing with lalrpop, as such I think moving to iterator is going to probably help more than hurt me

most of my usage of assumptions is only relying on ordering between variants, rather than
any zero based requirement  ensuring that the derive macro isn't reording them under the hood.

[rowan](https://github.com/rust-analyzer/rowan) if you look at its [examples](https://github.com/rust-analyzer/rowan/blob/master/examples/math.rs#L23) use a 0 based enum... maybe @matklad can better say why, maybe [here](https://github.com/rust-analyzer/rowan/blob/master/src/green/builder.rs#L88) it could be relying on the zero'ness, *not sure*.
I wonder if NonZeroU16 wouldn't somehow help things in memory layout for it.

If worst comes to worst I can subtract 1 when converting to rowan's Kind.
I was mainly thinking of interfacing with it, but haven't managed to for parser reasons.

The main way i've been using discriminator values myself is in translating between Token types when switching modes.

E.g. I'll have
```
#[derive(Logos,...)]
enum SubMode {
  #[end] 
  End = 0,
#[token = "\""]
  SomeToken = 1,
 #[regex = "..."]
 String = 2,
}

#[derive(Logos,...)]
enum TopMode {
  #[end] 
  End = 0,
 #[token = "\""]
  SomeToken = 1,
 // No Token/regex reserved for the SubMode
 String = 2,
}

One thing I wanted to ask, but failed to think of before the release was if it would be possible to reorder the discriminator constraint to allow Number of variants + 1 precisely for this optimization, and perhaps disallowing zero?

**Comment by maciejhirsz** [comment]
From what I can tell, `Option<Token>`, is the same as `Token` in memory for all discriminant values, that is if your last discriminant value is 3, then `None` is going to be 4, so converting `Option<T>` to `T` doesn't have to copy any memory or move any registers, it's the same byte.

> One thing I wanted to ask, but failed to think of before the release was if it would be possible to reorder the discriminator constraint to allow Number of variants + 1 precisely for this optimization, and perhaps disallowing zero?

I can relax the discriminant rules further, or even disable them altogether and drop the `lookup!` macro, because:
1. There doesn't seem to be that much use for it, Rust these days is really good at optimizing match expressions into jump tables.
2. I'm about to add support for enums with values, for which lookup tables wouldn't work at all.
So if you needed to manually reorder your discriminants to be 1, 2, 3..., I don't see why you couldn't.

**Comment by maciejhirsz** [comment]
Switched to `Option<Token>` internally and there is no discernable difference in benchmarks:

```
running 4 tests
test identifiers                       ... bench:         670 ns/iter (+/- 31) = 1162 MB/s
test keywords_operators_and_punctators ... bench:       1,974 ns/iter (+/- 113) = 1079 MB/s
test strings                           ... bench:         622 ns/iter (+/- 91) = 1400 MB/s
test x_identifiers_iter                ... bench:         698 ns/iter (+/- 38) = 1116 MB/s
```

Now I need to see if I can bring up iterator speed to the same as legacy loops and we'll be home.

**Comment by maciejhirsz** [comment]
I've just tried two different implementations for the iterator:
1. One that stores the token in a `token: Option<Token>` field inside the lexer, then removes it with `self.token.take()` when yielding `.next()`.
2. One that doesn't store the token internally at all and instead has the lexing process return `Option<Token>` directly.

Results are counter intuitive: the first option is faster, even as much as 10%. It's possible that tail call recursion is breaking down in the lexer at some points, in which case having to juggle return value on top of `&mut` lexer ref slows things down.

It was still a bit faster for me to use the `.advance()` mode, however once I modified my benchmarks to `black_box(token)` for each `Some(token)` found, it behave identical to my iterator benchmark, so I reckon there is no real drawback to embracing iterators.

**Comment by ratmice** [comment]
I've been looking at the lexer_modes test, and seeing if I couldn't implement a test for an iterator for bridging modes, similar to something i used to do, but haven't figured out where to_owned() went/came from :confused: 

in a fork of your [branch](https://github.com/ratmice/logos/tree/smarter-callbacks) here..

**Comment by maciejhirsz** [comment]
I don't recall ever having `to_owned` being a provided thing on `Logos` types. I _just_ pushed new changes that remove `.advance()` and makes `.token` a private field. `.advance_as()` is no longer a thing, and it's not necessary since `morph` is perfectly sufficient when using iterators:

https://github.com/maciejhirsz/logos/blob/8beda0970d08eb9fcd93353a75e567b8efe8d2b4/tests/tests/lexer_modes.rs#L38-L56

**Comment by ratmice** [comment]
Great, I must have missed the morph.

**Comment by maciejhirsz** [comment]
This is now done and released as 0.11.0-rc1 :)
